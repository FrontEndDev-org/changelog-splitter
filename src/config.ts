import fs from 'fs';
import path from 'path';
import { ExtendConfig, RuntimeConfig, StrictUserConfig, UserConfig } from './types';
import { createTempDirname, matchMajor } from './utils';

export const defaults: StrictUserConfig = {
  cwd: process.cwd(),
  currentChangelogFile: 'CHANGELOG.md',
  packageFile: 'package.json',
  previousVersionChangelogFileName: 'changelogs/v[major].x.md',
  previousVersionChangelogTitle: '# v[major].x 更新日志',
  previousVersionLinkTitle: '## 其他版本的更新日志',
  processVersion: () => 0,
};

/**
 * 定义配置
 * @param {UserConfig} config
 * @returns {StrictUserConfig}
 */
export function defineConfig(config?: UserConfig): StrictUserConfig {
  return Object.assign({}, defaults, config) as StrictUserConfig;
}

/**
 * 创建运行期配置
 * @param {StrictUserConfig} strictUserConfig
 * @returns {RuntimeConfig}
 */
export function createRuntimeConfig(strictUserConfig: StrictUserConfig): RuntimeConfig {
  const { cwd, packageFile, currentChangelogFile } = strictUserConfig;
  const resolvePath = (...segments: string[]): string => path.resolve(cwd, ...segments);
  const currentChangelogFilePath = resolvePath(currentChangelogFile);
  const currentMajorChangelogFilePath = path.join(createTempDirname(), `CHANGELOG.md`);

  const packageFilePath = resolvePath(packageFile);
  const json = fs.readFileSync(packageFilePath, 'utf8');
  const { version: currentVersion } = JSON.parse(json) as { version: string };
  const currentMajor = matchMajor(currentVersion);

  return Object.assign<any, any, ExtendConfig>({}, strictUserConfig, {
    resolvePath,
    currentChangelogFilePath,
    currentVersion,
    currentMajor,
    currentMajorChangelogFilePath,
  });
}
