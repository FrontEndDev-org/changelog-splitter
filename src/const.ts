export const pkgName = process.env.PKG_NAME!;
export const pkgVersion = process.env.PKG_VERSION!;

export const generatedSeparator = `<!-- Generated by ${pkgName}@v${pkgVersion}, please do not modify manually!! -->`;
export const generatedSeparatorRE = new RegExp(`<!-- .*?${pkgName}@v.*? -->`);
