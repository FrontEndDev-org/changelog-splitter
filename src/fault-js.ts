export type FaultDefinition = Record<string, string>;

// export interface Error

// export type ExtractFault =

export function defineFault<F extends FaultDefinition, C extends keyof F>(name: string, definition: F) {
  return class Fault extends Error {
    readonly name = name;
    readonly code: C;

    constructor(code: C, message?: string) {
      super(definition[code]);
      this.code = code;
      if (message) this.message = message;

      // @ref https://stackoverflow.com/a/32749533
      if (typeof Error.captureStackTrace === 'function') {
        Error.captureStackTrace(this, this.constructor);
      } else {
        this.stack = new Error(message).stack;
      }
    }

    because(cause: Error) {
      this.cause = cause;
      return this;
    }
  };
}
