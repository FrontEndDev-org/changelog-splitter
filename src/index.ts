export { run } from './commands';
export { defineConfig } from './config';
