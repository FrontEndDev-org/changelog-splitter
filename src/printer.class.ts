import chalk from 'chalk';
import path from 'path';

export class Printer {
  counts = {
    insertFiles: new Set<string>(),
    updateFiles: new Set<string>(),
    removeFiles: new Set<string>(),
  };

  static insertSymbol = chalk.cyanBright('+');
  static updateSymbol = chalk.yellowBright('~');
  static removeSymbol = chalk.redBright('-');

  constructor(private readonly cwd: string) {}

  private toRelative(to: string) {
    return path.relative(this.cwd, to);
  }

  insertFile(file: string) {
    if (this.counts.insertFiles.has(file)) return;

    console.log(Printer.insertSymbol, this.toRelative(file));
    this.counts.insertFiles.add(file);
  }

  updateFile(file: string) {
    if (this.counts.updateFiles.has(file)) return;

    console.log(Printer.updateSymbol, this.toRelative(file));
    this.counts.updateFiles.add(file);
  }

  removeFile(file: string) {
    if (this.counts.removeFiles.has(file)) return;

    console.log(Printer.removeSymbol, this.toRelative(file));
    this.counts.removeFiles.add(file);
  }
}
