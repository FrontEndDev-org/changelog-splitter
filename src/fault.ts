import { defineFault } from './fault-js';

const ConfigFaultDefinition = {
  CHANGELOG_NOT_FOUND: 'CHANGELOG 文件不存在',
  PACKAGE_JSON_READ_ERROR: 'package.json 文件读取失败',
  PACKAGE_JSON_PARSE_ERROR: 'package.json 文件解析失败',
};
export const ConfigFault = defineFault('ConfigFault', ConfigFaultDefinition);
export type ConfigFaultInstance = InstanceType<typeof ConfigFault>;
export type ConfigFaultCode = keyof typeof ConfigFaultDefinition;
