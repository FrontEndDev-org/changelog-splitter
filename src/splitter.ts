import * as fs from 'fs';
import * as path from 'path';
import { ConflictStrategy, createRuntimeConfig, RuntimeConfig, StrictUserConfig } from './config';
import { generatedSeparator, generatedSeparatorRE } from './const';
import { SplitFault } from './fault';
import {
  createTempFile,
  generateNameByMajor,
  matchPrevious,
  matchVersion,
  mergeFiles,
  pipeFile,
  readFileLineByLine,
} from './utils';

export interface SplitResult {
  runtimeConfig: RuntimeConfig;

  /**
   * 已经处理的主版本号与文件的映射表
   */
  processedFileByMajor: { [major: string]: string };

  /**
   * 主版本更新日志末尾的连续空白长度
   */
  blankLengthByMajor: { [major: string]: number };

  /**
   * 由于冲突策略导致的废弃文件
   */
  deprecatedMajorFiles: { [major: string]: string };
}

/**
 * 分离当前更新日志
 * @param {RuntimeConfig} runtimeConfig
 * @returns {Promise<void>}
 */
export async function splitCurrentChangelog(runtimeConfig: RuntimeConfig): Promise<SplitResult> {
  const {
    previousVersionChangelogTitle,
    previousVersionChangelogFileName,
    previousVersionChangelogConflictStrategy,
    resolvePath,
    currentChangelogFilePath,
    currentMajor,
    currentVersionChangeTempFilePath,
  } = runtimeConfig;

  const splitResult: SplitResult = {
    runtimeConfig,
    processedFileByMajor: {},
    blankLengthByMajor: {},
    deprecatedMajorFiles: {},
  };
  const { processedFileByMajor, blankLengthByMajor, deprecatedMajorFiles } = splitResult;

  const process = (major: string, line: string) => {
    const isCurrentMajor = major === currentMajor;
    const filePath = isCurrentMajor
      ? currentVersionChangeTempFilePath
      : resolvePath(generateNameByMajor(previousVersionChangelogFileName, major));

    fs.mkdirSync(path.dirname(filePath), { recursive: true });

    // 第一次处理，清空其本身内容
    if (!processedFileByMajor[major]) {
      fs.writeFileSync(filePath, '');
    }

    // 未处理过的大版本 && 不是当前版本 = 第一次处理旧版本更新日志
    if (!processedFileByMajor[major] && !isCurrentMajor) {
      fs.writeFileSync(filePath, `${generatedSeparator}\n\n`);

      // 旧版本标题
      if (previousVersionChangelogTitle) {
        const titleText = generateNameByMajor(previousVersionChangelogTitle, major);
        fs.appendFileSync(filePath, titleText + '\n\n');
      }
    }

    const blankLength = blankLengthByMajor[major] || 0;
    const isBlank = line.trim() === '';

    if (blankLength < 3 || !isBlank) {
      fs.appendFileSync(filePath, line + '\n');
    }

    blankLengthByMajor[major] = isBlank ? blankLength + 1 : 0;
    processedFileByMajor[major] = filePath;
  };

  let processingMajor = '';
  let processingPrevious = false;

  // title
  // version
  // ...
  // version
  // previousLink
  // ...
  // previousLink

  await readFileLineByLine(currentChangelogFilePath, async (line) => {
    const isPreviousBlock = generatedSeparatorRE.test(line);

    // 前版本块
    if (isPreviousBlock) {
      processingPrevious = true;
      return;
    }

    // 处理前版本引用链接，此时已经处理过更新日志中的所有版本了
    if (processingPrevious) {
      const previous = matchPrevious(line);

      if (!previous) return;

      const { link, major } = previous;
      // 旧文件
      const processedFile = resolvePath(link);
      // 新文件
      const processingFile =
        processedFileByMajor[major] || resolvePath(generateNameByMajor(previousVersionChangelogFileName, major));
      // 是否存在两份旧版本文件
      const conflicting = processedFile !== processingFile;

      if (!conflicting) return;

      const existProcessingFile = fs.existsSync(processingFile);
      const existProcessedFile = fs.existsSync(processedFile);

      // 新旧文件同时存在：合并文件
      const acceptedFile =
        previousVersionChangelogConflictStrategy === ConflictStrategy.ProcessedFile
          ? (processedFileByMajor[major] = processedFile)
          : (processedFileByMajor[major] = processingFile);

      if (existProcessingFile && existProcessedFile) {
        const tempFile = createTempFile();
        await mergeFiles(
          [
            // 旧文件在前
            processedFile,
            // 新文件在后
            processingFile,
          ],
          tempFile
        );
        await pipeFile(tempFile, acceptedFile);
        if (processedFile !== acceptedFile) deprecatedMajorFiles[major] = processedFile;
      }
      // 只有旧文件
      else if (existProcessedFile) {
        await pipeFile(processedFile, acceptedFile);
        if (processedFile !== acceptedFile) deprecatedMajorFiles[major] = processedFile;
      } else {
        throw new SplitFault('LINK_CHANGELOG_NOT_FOUND', `链接的更新日志文件不存在 ${processedFile}`);
      }

      return;
    }

    const version = matchVersion(line);

    // 版本开始
    if (version) {
      processingMajor = version.major;
      process(processingMajor, line);
    }
    // 版本区
    else if (processingMajor) {
      process(processingMajor, line);
    }
    // 标题块
    else {
      process(currentMajor, line);
    }
  });

  return splitResult;
}

/**
 * 引用之前的更新日志链接
 * @param {RuntimeConfig} runtimeConfig
 * @param {SplitResult} splitResult
 */
export async function referPreviousChangelog(runtimeConfig: RuntimeConfig, splitResult: SplitResult) {
  const { currentVersionChangeTempFilePath, currentMajor, previousVersionLinkTitle, currentVersionChangeFilePath } =
    runtimeConfig;

  const { processedFileByMajor, blankLengthByMajor } = splitResult;
  const prevVersions = Object.keys(processedFileByMajor)
    .filter((v) => v !== currentMajor)
    .map((v) => parseInt(v, 10))
    .sort((a, b) => b - a);

  // 需要链接其他版本
  if (prevVersions.length > 0) {
    // 通常是存在的，为了便于单元测试时不存在
    if (!fs.existsSync(currentVersionChangeTempFilePath)) {
      fs.mkdirSync(path.dirname(currentVersionChangeTempFilePath), { recursive: true });
    }

    if (!blankLengthByMajor[currentMajor]) {
      fs.appendFileSync(currentVersionChangeTempFilePath, `\n\n`);
    }

    fs.appendFileSync(currentVersionChangeTempFilePath, `${generatedSeparator}\n\n`);
    fs.appendFileSync(currentVersionChangeTempFilePath, `${previousVersionLinkTitle}\n`);

    const currentDir = path.dirname(currentVersionChangeFilePath);
    prevVersions.forEach((v) => {
      const filePath = processedFileByMajor[v];
      const relativePath = path.relative(currentDir, filePath);
      fs.appendFileSync(currentVersionChangeTempFilePath, `- [v${v}.x](${relativePath})\n`);
    });

    fs.appendFileSync(currentVersionChangeTempFilePath, '\n');
  }

  // 通常是存在的，为了便于单元测试时不存在
  if (fs.existsSync(currentVersionChangeTempFilePath)) {
    // 复制当前大版本的更新日志临时文件回原地
    await pipeFile(currentVersionChangeTempFilePath, currentVersionChangeFilePath);
    fs.rmSync(currentVersionChangeTempFilePath);
  }
}

export async function splitChangelog(config: StrictUserConfig) {
  const runtimeConfig = createRuntimeConfig(config);
  const result = await splitCurrentChangelog(runtimeConfig);
  await referPreviousChangelog(runtimeConfig, result);
  return result;
}
