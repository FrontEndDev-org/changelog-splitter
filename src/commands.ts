import chalk from 'chalk';
import { SingleBar } from 'cli-progress';
import { tryFlatten } from 'try-flatten';
import { defineConfig, UserConfig } from './config';
import { Printer } from './printer.class';
import { splitChangelog, SplitProcessingStage, SplitResult } from './splitter';

export function printResult(splitResult: SplitResult) {
  const { splitContext, runtimeConfig } = splitResult;
  const { cwd, currentVersionChangeFilePath, currentChangelogFilePath, currentVersionChangeTempFilePath } =
    runtimeConfig;
  const { processedFileByMajor, deprecatedMajorFiles } = splitContext;
  const printer = new Printer(cwd);

  console.log(`更新日志文件变化情况如下（删除标记“${Printer.removeSymbol}”的文件需要手动删除）：`);

  // current version changelog
  if (currentChangelogFilePath === currentVersionChangeFilePath) {
    printer.updateFile(currentVersionChangeFilePath);
  } else {
    printer.removeFile(currentChangelogFilePath);
    printer.insertFile(currentVersionChangeFilePath);
  }

  // previous version changelog
  for (const [major, file] of Object.entries(processedFileByMajor)) {
    if (file === currentVersionChangeTempFilePath) continue;
    printer.insertFile(file);
  }
  for (const [major, file] of Object.entries(deprecatedMajorFiles)) {
    printer.removeFile(file);
  }
}

export async function run(userConfig: UserConfig) {
  const strictUserConfig = defineConfig(userConfig);
  const stageNames: Record<SplitProcessingStage, string> = {
    [SplitProcessingStage.Parse]: '解析',
    [SplitProcessingStage.Refer]: '引用',
  };
  const createBar = (stage: SplitProcessingStage) => {
    const bar = new SingleBar({
      format: `${stageNames[stage]} [{bar}] {percentage}%`,
    });
    bar.start(100, 0);
    return bar;
  };
  const bars: Record<SplitProcessingStage, SingleBar | null> = {
    [SplitProcessingStage.Parse]: null,
    [SplitProcessingStage.Refer]: null,
  };

  console.log('更新日志文件切割进行中...');
  console.log();

  const [err, res] = await tryFlatten(
    splitChangelog(strictUserConfig, (processing) => {
      const bar = (bars[processing.stage] = bars[processing.stage] || createBar(processing.stage));
      bar.update(processing.progress * 100);
      if (processing.progress === 1) bar.stop();
    })
  );

  console.log();

  if (err) {
    console.log(chalk.redBright('更新日志文件切割失败'));
    console.log(chalk.redBright(err.message));
    return;
  }

  printResult(res);
  console.log();
  console.log(chalk.greenBright('更新日志文件切割成功'));
}
