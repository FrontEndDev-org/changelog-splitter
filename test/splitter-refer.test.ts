import fs from 'fs';
import { expect, test } from 'vitest';
import { defineConfig } from '../src';
import { createRuntimeConfig } from '../src/config';
import { createSplitContext, referPreviousChangelog, SplitContext } from '../src/splitter';
import { ChangelogFolder, makeChangelogCwd } from './helpers';

test(ChangelogFolder.EmptyVersions, async () => {
  const [cwd, clean] = makeChangelogCwd(ChangelogFolder.EmptyVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const { currentVersionChangeFilePath } = config;
  const splitContext = createSplitContext();

  await referPreviousChangelog(config, splitContext);
  expect(fs.readFileSync(currentVersionChangeFilePath, 'utf8')).toMatchSnapshot();

  clean();
});

test(ChangelogFolder.OnlyCurrentVersion, async () => {
  const [cwd, clean] = makeChangelogCwd(ChangelogFolder.OnlyCurrentVersion);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const { currentMajor, resolvePath, currentVersionChangeFilePath } = config;
  const splitContext = Object.assign(createSplitContext(), {
    processedFileByMajor: {
      [currentMajor]: resolvePath('123.md'),
    },
  });
  await referPreviousChangelog(config, splitContext);
  // 空
  expect(fs.readFileSync(currentVersionChangeFilePath, 'utf8')).toMatchSnapshot();

  clean();
});

test(ChangelogFolder.NotCurrentVersion, async () => {
  const [cwd, clean] = makeChangelogCwd(ChangelogFolder.NotCurrentVersion);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const { resolvePath, currentVersionChangeFilePath } = config;
  const splitContext = Object.assign(createSplitContext(), {
    processedFileByMajor: {
      2: resolvePath('123.md'),
    },
  });

  await referPreviousChangelog(config, splitContext);
  // 1 个引用链接
  expect(fs.readFileSync(currentVersionChangeFilePath, 'utf8')).toMatchSnapshot();

  clean();
});

test(ChangelogFolder.HasManyVersions, async () => {
  const [cwd, clean] = makeChangelogCwd(ChangelogFolder.HasManyVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const { resolvePath, currentVersionChangeFilePath } = config;
  const splitContext = Object.assign(createSplitContext(), {
    processedFileByMajor: {
      2: resolvePath('123.md'),
      14: resolvePath('456.md'),
      16: resolvePath('789.md'),
    },
  });
  await referPreviousChangelog(config, splitContext);
  // 3 个引用链接
  expect(fs.readFileSync(currentVersionChangeFilePath, 'utf8')).toMatchSnapshot();

  clean();
});

test(ChangelogFolder.HasRefVersions, async () => {
  const [cwd, clean] = makeChangelogCwd(ChangelogFolder.HasRefVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const { resolvePath, currentVersionChangeFilePath } = config;
  const splitContext = Object.assign(createSplitContext(), {
    processedFileByMajor: {
      2: resolvePath('123.md'),
      14: resolvePath('456.md'),
      16: resolvePath('789.md'),
    },
  });
  await referPreviousChangelog(config, splitContext);
  // 3 个引用链接
  expect(fs.readFileSync(currentVersionChangeFilePath, 'utf8')).toMatchSnapshot();

  clean();
});
