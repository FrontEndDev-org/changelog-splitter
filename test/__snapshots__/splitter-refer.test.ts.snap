// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`1-empty-versions 1`] = `""`;

exports[`2-only-current-version 1`] = `
"# Changelog

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)
"
`;

exports[`3-not-current-version 1`] = `
"

<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

## 其他版本的更新日志
- [v2.x](123.md)

"
`;

exports[`4-has-many-versions 1`] = `
"

<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

## 其他版本的更新日志
- [v14.x](456.md)
- [v2.x](123.md)

"
`;

exports[`5-has-ref-versions 1`] = `
"

<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

## 其他版本的更新日志
- [v14.x](456.md)
- [v2.x](123.md)

"
`;

exports[`referPreviousChangelog > 1-empty-versions 1`] = `""`;

exports[`referPreviousChangelog > 2-only-current-version 1`] = `
"# Changelog

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)
"
`;

exports[`referPreviousChangelog > 3-not-current-version 1`] = `
"

<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

## 其他版本的更新日志
- [v2.x](123.md)

"
`;

exports[`referPreviousChangelog > 4-has-many-versions 1`] = `
"

<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

## 其他版本的更新日志
- [v14.x](456.md)
- [v2.x](123.md)

"
`;
